### PR TITLE
[AMD] Fix custom AR capture for DSA DP dual graphs

### LINKED ISSUE #36071: [Bug][AMD] Custom AR graph registration corrupts DSA DP-attention graph-skew replay
## Summary

On ROCm, custom all-reduce direct graph-input registration is unsafe when DSA
DP-attention lanes independently replay different decode graphs. On GLM-5.2
TP8/DP4 this can make the collective incomplete, turn every output logit into
NaN, and emit token ID 0 repeatedly (`!!!!!!!!`).

The failure does not require MTP, overlap scheduling, AITER prefill, or AITER
all-reduce fusion. Eager decode, plain NCCL, and the custom-AR stable copy-in
path are clean under the same model and parallelism.

## Environment

- 8 x AMD MI350X (`gfx950`)
- ROCm 7.2.4
- SGLang revision `db570fe619ae15d8bd4ee04f146bc8234c52a20d`
- AITER revision `c16d44b93a528b2a4bfd6d8d3409116d465872a9`
- Model: `amd/GLM-5.2-MXFP4`, real weights
- KV cache: BF16
- TP8, DP4 with DP attention, EP8
- TileLang DSA prefill and decode

## Minimal server configuration

```bash
SGLANG_USE_AITER=1 \
SGLANG_USE_ROCM700A=1 \
SGLANG_ROCM_FUSED_DECODE_MLA=1 \
python3 -u -m sglang.launch_server \
  --model-path amd/GLM-5.2-MXFP4 \
  --served-model-name glm-5.2 \
  --tp-size 8 \
  --dp-size 4 \
  --enable-dp-attention \
  --ep-size 8 \
  --moe-a2a-backend none \
  --kv-cache-dtype bf16 \
  --mem-fraction-static 0.85 \
  --attention-backend dsa \
  --dsa-prefill-backend tilelang \
  --dsa-decode-backend tilelang \
  --dsa-topk-backend sgl-kernel \
  --enable-aiter-allreduce-fusion \
  --disable-overlap-schedule \
  --cuda-graph-backend-decode full \
  --cuda-graph-max-bs-decode 128 \
  --cuda-graph-bs-decode 1 2 4 8 16 32 64 128 \
  --cuda-graph-backend-prefill disabled \
  --quantization quark \
  --trust-remote-code
```

MTP is intentionally disabled in the focused reproducer.

## Actual result

Decode eventually emits token ID 0 with null logprobs:

```text
We need to write a correct Python function!!!!!!!!

output_ids:
[1654, 1184, 311, 3270, 264, 4396, 13020, 729, 0, 0, 0, 0, 0, 0, 0, 0]
```

Both attention-TP ranks in the routed DP lane report all-NaN logits:

```text
shape=(1, 154880) finite=0 nan=154880
```

## Root cause

This is not a DSA attention-kernel error or a shared graph-memory-pool pointer
collision.

During capture, AITER/SGLang direct custom-AR registration associates every
captured call with a `RankData` slot. Post-capture registration fills that slot
with peer pointers from the same captured-call ordinal. Replay has no runtime
graph-id handshake and therefore assumes every communicator rank is replaying
the same captured graph/call ordinal.

DSA DP-attention lanes can select different graphs concurrently. A TP2 pair
replaying graph 0 consequently reads graph-0 buffers on peers currently
producing graph 1, 2, or 3. Those buffers are stale/inactive and the global
collective is incomplete.

## Minimal causal proof

A standalone eight-GPU test captures four same-shape BF16 graphs with unique
per-graph values:

| Input path | Replay pattern | Failed rank-results |
|---|---|---:|
| direct registration | all ranks replay the same graph | 0/32 |
| direct registration | DP4/TP2 graph skew | 8/8 |
| stable copy-in | identical graph skew | 0/8 |

Expected global sum: `2080` on every rank. Direct/skew mode returns `206`,
`414`, `624`, and `832`, exactly the four active TP2-pair contributions.
Copy-in returns `2080` on all eight ranks.

Instrumentation of 126,416 capture/registration events found no registration
count, IPC-handle/offset, or capture-to-registration drift. A stronger
release/acquire system-scope barrier still failed 8/8 full-model requests.

## Full-model isolation

| Collective path | OSL32 result |
|---|---:|
| AITER custom AR, direct graph registration | 4/4 corrupt |
| SGLang custom AR, direct graph registration | 4/4 corrupt |
| Plain NCCL | 0/4 corrupt |
| AITER custom AR, stable copy-in | 0/4 corrupt |

Copy-in subsequently passed 32/32 sequential and 32/32 concurrent requests,
including the production MTP + overlap configuration.

## Fix

#36088 retains custom AR but uses its stable pre-registered copy-in buffer
during graph capture for `DSA + DP attention`. This removes graph identity from
the peer-pointer contract. The measured decode throughput delta was -0.40%.

A future zero-copy design would need runtime active-graph-slot indirection or
allocator integration that makes every graph produce into the same stable
registered input address.

### PR DESCRIPTION
**Pull request by tanth47**
## Motivation

Fixes #36071.

On ROCm, DSA with DP attention allows different DP lanes to replay different
decode graphs concurrently. Custom all-reduce direct graph-input registration
assumes something stronger: every rank in the communicator replays the same
captured graph and custom-AR call ordinal in lockstep.

During capture, each custom-AR call embeds a `RankData` slot. Registration then
fills that slot with peer pointers belonging to the same captured-call ordinal.
At replay there is no runtime graph-id handshake. If TP2 lane pairs select
different graphs, each pair reads graph-local buffers that are inactive on the
other ranks. The resulting all-reduce is incomplete and can turn GLM-5.2 logits
into NaN, followed by repeated token ID 0.

This supersedes the earlier shared-memory-pool address-collision hypothesis.
Instrumentation found no pointer-registration ordering, IPC-handle, offset, or
capture-to-registration drift.

## Modifications

- Use the communicator's stable pre-registered copy-in buffer during graph
  capture when DSA DP attention is enabled.
- Apply the policy to both AITER and legacy SGLang custom all-reduce.
- Preserve direct graph-input registration for non-DP DSA and other attention
  backends.
- Add unit coverage for the graph-capture policy.

This retains custom-AR itself. Only its graph input staging changes.

## Minimal causal proof

Hardware: 8 x AMD MI350X, ROCm 7.2.4, AITER
`c16d44b93a528b2a4bfd6d8d3409116d465872a9`.

A standalone distributed reproducer captures four same-shape BF16 graphs. Each
graph contributes a distinct constant, making stale peer-buffer selection
numerically observable.

| Input registration | Replay pattern | Failed rank-results | Result |
|---|---|---:|---|
| direct | all eight ranks replay the same graph | 0/32 | exact |
| direct | DP4/TP2 skew: rank pairs replay g0/g1/g2/g3 | 8/8 | incorrect |
| copy-in | the same DP4/TP2 skew | 0/8 | exact |

The expected skewed global sum is BF16 `2080` on every rank. Direct mode
returns `206`, `414`, `624`, and `832` for the four rank pairs. These are
exactly the contributions from each active TP2 pair, proving that the other
graphs' peer buffers were not supplying the current inputs. Copy-in returns
`2080` on all eight ranks.

A stronger release/acquire system-scope barrier did not help (8/8 full-model
requests still corrupt), excluding a simple memory-visibility issue.

## Full-model accuracy tests

Model/configuration: GLM-5.2-MXFP4, BF16 KV, TP8/DP4/EP8 with DP attention,
TileLang DSA prefill/decode, full decode graphs.

| Graph collective | Result |
|---|---:|
| AITER custom AR, direct registration | 4/4 hard corrupt |
| SGLang custom AR, direct registration | 4/4 hard corrupt |
| Plain NCCL | 0/4 hard corrupt |
| AITER custom AR, copy-in | 0/4 hard corrupt |

Extended copy-in soak:

- no MTP: sequential 32/32 clean; concurrent c8 x 4 OSL64 32/32 clean;
- AITER prefill + TileLang decode + MTP + overlap: sequential 32/32 clean;
  concurrent c8 x 4 OSL64 32/32 clean;
- six policy/dispatch unit tests pass in the ROCm image.

## Performance

Three matching rounds with AITER prefill + MTP, c4, 4,096 cached + 1,024
fresh, OSL512:

| Capture input path | Mean TPOT | Mean token/s/user |
|---|---:|---:|
| direct registration | 11.753 ms | 85.14 |
| copy-in | 11.889 ms | 84.80 |

Observed throughput delta: **-0.40%**, within run-to-run variance from MTP
acceptance length.

## Design scope

The patch is the smallest safe production fix. A zero-copy replacement would
require a new runtime protocol that advertises each rank's active graph slot,
or allocator integration that makes every graph produce directly into one
stable registered address. Static same-ordinal peer registration cannot safely
support independently selected graphs.

## Checklist

- [x] Format code according to project style.
- [x] Add unit tests.
- [x] Provide accuracy and speed benchmark results.
- [x] Follow the SGLang code style guidance.

---
### CI States

Latest PR Test (Base): :x: [Run #33102231789](https://github.com/sgl-project/sglang/actions/runs/33102231789)
Latest PR Test (Extra): :x: [Run #33102231463](https://github.com/sgl-project/sglang/actions/runs/33102231463)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33102231608](https://github.com/sgl-project/sglang/actions/runs/33102231608)

### CODE REVIEW

**Comment by tanth47** [comment]
@HaiShaw, could you take a look? Thanks!

**Comment by hubertlu-tw** [comment]
This PR adds `enable_register_for_capturing` to the shared `CustomAllreduce` constructor and uses that flag in the capture path, so CUDA/MUSA files are in the diff. Behavior there is unchanged: CUDA/MUSA still return the class from `dispatch_custom_allreduce` without this policy, so the flag stays at its default (True unless memory-saver CUDA graph is on, which already used copy-in).

The new copy-in path is HIP-only and only when DP attention is on and the decode attention backend is `dsa`. DSA without DP attention, DP attention without DSA, and every other backend keep direct graph-input registration.
